src/constraint/CollisionFree.cpp

```diff
@@ -42,16 +43,47 @@ bool CollisionFree::isSatisfied(
         groups.second.get(),
         mCollisionOptions,
         &collisionResult);
+
     if (collision)
+    {
+      if (_outcome != nullptr)
+      {
+        const auto& collisionOutcome
+            = dynamic_cast<CollisionFreeOutcome*>(_outcome);
```

Comment:
Nit: `dynamic_cast` is known to be expensive (relatively to `static_cast`) so it's preferable to avoid or use less if possible. According to this logic, dynamic casting will be called multiple times. I think this can be optimized by performing the dynamic casting once like:
```c++
CollisionFreeOutcome* collisionOutcome = nullptr;
  if (_outcome)
  {
    collisionOutcome = dynamic_cast<CollisionFreeOutcome*>(_outcome);
    if (!collisionOutcome)
    {
      throw std::invalid_argument(
          "TestableOutcome pointer is not of type CollisionFreeOutcome.");
    }
  }

  for (const auto& groups : mGroupsToPairwiseCheck)
  {
    collision = mCollisionDetector->collide(
        groups.first.get(),
        groups.second.get(),
        mCollisionOptions,
        &collisionResult);

    if (collision)
    {
      if (collisionOutcome)
      {
        const auto& contacts = collisionResult.getContacts();
        for (const auto& elem : contacts)
          collisionOutcome->markPairwiseContact(elem);
      }

      return false;
    }
  }

  for (const auto& group : mGroupsToSelfCheck)
  {
    collision = mCollisionDetector->collide(
        group.get(), mCollisionOptions, &collisionResult);
    
    if (collision)
    {
        if (collisionOutcome)
        {
          const auto& contacts = collisionResult.getContacts();
          for (const auto& elem : contacts)
            collisionOutcome->markSelfContact(elem);
        }

      return false;
    }
  }

  return true;
```

Reply:
I initially had the same thought, but I believe that since we immediately `return false` when there is a pairwise collision, there is no chance of `dynamic_cast`ing a second time.

Reply:
What if `outcome` is actually a pointer of `CollisionFreeOutcome`? I think in that case, `dynamic_cast` will be called multiple times for every iteration of the pairwise collisions.